### PR TITLE
Send order lines in orderGrantRefundCreate mutation

### PR DESCRIPTION
**Pull request by poulch**
closes https://github.com/saleor/saleor-dashboard/issues/4277

### General AC:
- A user should be able to choose each line quantity, is shipping refunded, refund amount and send those data to API
- A user can choose shipping refund only when none of the other refunds contains a shipping refund selected
-A user can only select the quantity that is able to refund, when the user had ordered an item with quantity 3, refund 2 of 3, now can only refund 1
- A user during the creation of a grant refund should see the calculated refund amount based on the selected quantity and shipping refund. User can also provide own custom refund amount.
- A user during editing granted refund should always see the refunded amount, changing the quantity and shipping refund does not influence the refund amount. User can manually update refund amount.

### Screenshots

Before
Screenshot 2023-10-30 at 10 46 14

After
Screenshot 2023-10-30 at 10 48 18

### Pull Request Checklist

1. [ ] This code contains UI changes
2. [ ] All visible strings are translated with proper context including data-formatting
3. [ ] Attributes `data-test-id` are added for new elements
4. [ ] The changes are tested in Chrome/Firefox/Safari browsers and in light/dark mode
5. [ ] Your code works with the latest stable version of the core
6. [ ] I added changesets file (instructions in [contribution guide](https://github.com/saleor/saleor-dashboard/blob/main/.github/CONTRIBUTING.md)

### Test environment config

API_URI=https://master.staging.saleor.cloud/graphql/
APPS_MARKETPLACE_API_URI=https://apps.staging.saleor.io/api/v2/saleor-apps

### Do you want to run more stable tests?

To run all tests, just select the stable checkbox. To speed up tests, increase the number of containers. Tests will be re-run only when the "run e2e" label is added.

1. [ ] stable
2. [ ] app
3. [ ] attribute
4. [ ] category
5. [ ] collection
6. [ ] customer
7. [ ] giftCard
8. [ ] homePage
9. [ ] login
10. [ ] menuNavigation
11. [ ] navigation
12. [ ] orders
13. [ ] pages
14. [ ] payments
15. [ ] permissions
16. [ ] plugins
17. [ ] productType
18. [ ] products
19. [ ] sales
20. [ ] shipping
21. [ ] translations
22. [ ] variants
23. [ ] vouchers

CONTAINERS=6

### CODE REVIEW

File: src/orders/components/OrderGrantRefundPage/OrderGrantRefundPage.tsx [hunk 1]
```diff
@@ -103,81 +148,118 @@ const OrderGrantRefundPage: React.FC<OrderGrantRefundPageProps> = ({
...
+                      isFormDirty
+                        ? amountValue.toString()
+                        : formatMoneyAmount(
+                            {
+                              amount: amountValue,
+                              currency,
+                            },
+                            locale,
+                          )
+                    }
```
Comment (andrzejewsky) [inline src/orders/components/OrderGrantRefundPage/OrderGrantRefundPage.tsx:243]: suggestion: that's good candidate for separate function

File: src/orders/components/OrderGrantRefundPage/reducer.ts [hunk 2]
```diff
@@ -20,47 +27,85 @@ export type GrantRefundAction =
...
+            line.quantity,
+            order?.grantedRefunds,
+            grantedRefund?.id,
+          ),
+          unitPrice: line.unitPrice.gross.amount,
+          selectedQuantity: initialQuantity,
+          initialQuantity,
+        },
+      ];
+    });
```
Comment (andrzejewsky) [inline src/orders/components/OrderGrantRefundPage/reducer.ts:79]: suggestion: this entire map function can be lifted up and named accordingly to what it does

File: src/orders/components/OrderGrantRefundPage/reducer.ts [hunk 3]
```diff
@@ -20,47 +27,85 @@ export type GrantRefundAction =
...
+            line.quantity,
+            order?.grantedRefunds,
+            grantedRefund?.id,
+          ),
+          unitPrice: line.orderLine.unitPrice.gross.amount,
+          selectedQuantity: initialQuantity,
+          initialQuantity,
+        },
+      ];
+    });
```
Comment (andrzejewsky) [inline src/orders/components/OrderGrantRefundPage/reducer.ts:104]: issue: this almost completely copy pasted, the only difference is `line.orderLine.unitPrice.gross.amount`

**Review by andrzejewsky — CHANGES_REQUESTED** [review]

**Review by andrzejewsky — APPROVED** [review]

**Comment by witoszekdev** [comment]
@poulch The `OrderGrantRefundCreateLineInput` has a `reason` field as well, aka you can provide a reason for each field separately, like this:

```json
{
  "lines": [{"reason": "..." }],
  "reason": ""
}
```

Will we support this in the next PR, or am I missing something?

**Comment by poulch** [comment]
> @poulch The `OrderGrantRefundCreateLineInput` has a `reason` field as well, aka you can provide a reason for each field separately, like this:
> 
> ```json
> {
>   "lines": [{"reason": "..." }],
>   "reason": ""
> }
> ```
> 
> Will we support this in the next PR, or am I missing something?

Currently, the UI does not allow the user to add a reason to each line. However, we can address this issue separately since it requires UI redesign.

**Review by Droniu — APPROVED** [review]

**Review by witoszekdev — APPROVED** [review]

**Review by witoszekdev — DISMISSED** [review]

**Comment by wojteknowacki** [comment]
Adjust currency marks in refund amount input
Zrzut ekranu 2023-10-31 o 14 28 21

**Comment by wojteknowacki** [comment]
Switching back the refund toggle should be enough to save the edited refund


https://github.com/saleor/saleor-dashboard/assets/124166231/cd6acdfe-4432-4397-a72d-12fc75c31cd6

**Comment by wojteknowacki** [comment]
Error when refunding only shipping
Zrzut ekranu 2023-10-31 o 14 56 00

**Review by wojteknowacki — APPROVED** [review]

**Review by Droniu — APPROVED** [review]

**Pull request merged**